In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [98]:
df = pd.read_csv('https://github.com/WalePhenomenon/climate_change/blob/master/fuel_ferc1.csv?raw=true',error_bad_lines=False)
df.to_csv('fuel_data.csv',index=False)
#this is used to get data from the web and also converted to a local .csv file which will be saved as 'fuel_data'

In [99]:
#reading the fuel_data.csv file and making it a dataframe to be used for the quiz
df = pd.read_csv('fuel_data.csv')

# QUESTION 1

In [96]:
#question 2
A = [1,2,3,4,5,6]
B = [13,21,34]

In [95]:
A.append(B)
print (A)
#this will add B to the end of the list as a single element

[1, 2, 3, 4, 5, 6, [13, 21, 34]]


In [97]:
A.extend(B)
print (A)
#this add each element of B to the list not as a single element

[1, 2, 3, 4, 5, 6, 13, 21, 34]


# THIS MAKES OPTION B THE ANSWER

# QUESTION 2

In [5]:

np.identity(3)
# this will create an identity matrix of 3 by 3 with diagonal element of 1 and other elements to be zeros 


array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [93]:
np.array[(1,0,0),(0,1,0),(0,0,1)]
np.array([1,0,0],[0,1,0],[0,0,1])
#both values above does not give am identity matrix but can be modifed to form an identity matrix

TypeError: 'builtin_function_or_method' object is not subscriptable

# QUESTION 3

In [80]:
#question 3
df.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean()
#this uses groupby function, which is to group the fuel type and mmbtu per unit


fuel_type_code_pudl
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: fuel_cost_per_unit_burned, dtype: float64

from the above answer it is seen that waste has the lowest average fuel cost per unit burned

# QUESTION 4

In [21]:
df['fuel_mmbtu_per_unit'].describe()
#this gives the summary statistics of the feature selected


count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

fromt the above description , the standard deviation and the 75th percentile of energy per unit are 10.60 and 17.60

# QUESTION 5

In [84]:
df['fuel_qty_burned'].skew()

#to get skewness for thee fuel quantity burned , we use the function .skew()


15.851495469109503

In [27]:
df['fuel_qty_burned'].kurt()
#to get the kurtosis for thee fuel quantity burned , we use the function .kurt()


651.3694501337732

# QUESTION 6

In [89]:
missing_value = df.isnull().sum()
missing_value

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

it is seen here that fuel_unit has 180  missng values 

In [87]:
df2= df['fuel_unit'].count()
df2
#to get the total number of fuel_unit

29343

In [90]:
(missing_value/df2)*100
#to get the percentage of the missing value which is (missing vlaue/total number)*100

record_id                       0.000000
utility_id_ferc1                0.000000
report_year                     0.000000
plant_name_ferc1                0.000000
fuel_type_code_pudl             0.000000
fuel_unit                       0.613434
fuel_qty_burned                 0.000000
fuel_mmbtu_per_unit             0.000000
fuel_cost_per_unit_burned       0.000000
fuel_cost_per_unit_delivered    0.000000
fuel_cost_per_mmbtu             0.000000
dtype: float64

# QUESTION 7

Having checked through the data of 'fuel_data', it is seen that the column compose of strings values that are identitical which implies they are not numerical i.e they are categorical and such can only be replaced with the highest occuring value which is the mode 

In [41]:
df['fuel_unit'].value_counts()


mcf       11354
ton        8958
bbl        7998
gramsU      464
mmbtu       180
kgU         110
mwhth       100
mwdth        95
gal          84
Name: fuel_unit, dtype: int64

the highest value here is mcf, so the missing values can be replaced with mcf which is categorical

# QUESTION 8

In [42]:
df.corr()

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

from the correlation table shown above, fuel_mmbtu_per_unit and fuel_qty_burned have the third and second lowest correlation with fuel_cost_per_unit_burned

# QUESTION 9 

In [70]:
df1= df.groupby(['fuel_type_code_pudl','report_year'])['fuel_cost_per_unit_burned'].sum().reset_index()
df1
#this will group the values of report year and fuel type code by fuel burned using index for easy identification or calling out a value

fuel_type_code_pudl  report_year  fuel_cost_per_unit_burned
0                  coal         1994                  14984.572
1                  coal         1995                  14571.785
2                  coal         1996                  12694.803
3                  coal         1997                  43742.178
4                  coal         1998                  11902.597
..                  ...          ...                        ...
143               waste         2014                    591.594
144               waste         2015                    629.270
145               waste         2016                    539.079
146               waste         2017                    452.079
147               waste         2018                    660.972

[148 rows x 3 columns]

In [73]:
year_1994= df1['fuel_cost_per_unit_burned'][0]
#year 1994 is the first row under fuel_cost_per_unit_burned
year_1998 = df1['fuel_cost_per_unit_burned'][4]
#year 1998 is the first row under fuel_cost_per_unit_burned


In [75]:
((year_1998 - year_1994)/year_1994)*100

-20.56765451826049

The percentage change in the fuel cost per uit burned in 1998 compared to 1994 is -21% approximately

# QUESTION 10

In [63]:
fuel_cost_year =  df['fuel_cost_per_unit_delivered'].groupby(df['report_year'])
#to group fuel_cost_per_unit_delivered by report_year

In [62]:
average_fuel_cost_per_unit_delivered = fuel_cost_year.mean()
#to find the aveage of the grouped data

In [65]:
average_fuel_cost_per_unit_delivered.sort_values(ascending = False)
#to print out the average values in descending order

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: fuel_cost_per_unit_delivered, dtype: float64

From the table above it is shown that 1997 has the highest average fuel cost per unit delivered